src/osd/pg_scrubber.h

```diff
@@ -0,0 +1,631 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <cassert>
+#include <chrono>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "PG.h"
+#include "ScrubStore.h"
+#include "scrub_machine_lstnr.h"
+#include "scrubber_common.h"
+
+class Callback;
+
+namespace Scrub {
+class ScrubMachine;
+struct BuildMap;
+
+
+/**
+  ReplicaReservations is an RAII wrapper of the process of reserving/freeing scrub
+  resources at the replicas.
+
+  When constructed - sends reservation requests to the acting_set.
+  Handles arriving responses:
+  - a 'grant' is marked by reducing the 'pending_' counter. A list of reserved
+    resources is maintained.
+  - a rejection triggers a "couldn't acquire the replicas' scrub resources" event.
+    All "grants" arriving after the first rejection are denied, and cause an immediate
+    release of the resource.
+
+  Upon destruction - the acting-set (actually - those of the acting-set that have
+  'granted' our request - are freed.
+*/
+class ReplicaReservations {
+  using OrigSet =
+    decltype(std::declval<PG>().get_actingset());  // as I plan to change this type
+  using Vec = std::vector<pg_shard_t>;	// to be replaced with small_vector or small_set
+
+  PG* pg_;  // a pointer, as to not affect ReplicaReservations 'triviality'
```

Comment:
No, it's a pointer because this class does not "own" the lifecycle of PG.  The user is responsible for ensuring that PG outlives any ReplicaReservation instance.

Reply:
The comment meant to explain why I am not using a ref